### PR TITLE
Import/Export mapping by json file

### PR DESCRIPTION
**Pull request by anhnq-1924**
## Related Tickets

- [#520610](https://dev.sun-asterisk.com/issues/520610)

## What's this PR do ?

- [x] Implement function A
- [x] Fix bugs on function B

## Library
*(List package, library third party add new include version)*

- Library: google-analysis.js

## Performance

- [ ] Resolved n + 1 query
- [ ] Run explain query already
- [ ] Time run rake task : 1000 ms

## Checklist

- [ ] It was tested in local success?
- [ ] Fill link PR into ticket and the opposite
- [ ] Note purpose/reason, solution, scope of influence into ticket

## ENV note

```
```

## Notes
*(Other notes)*

### CODE REVIEW

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>failure</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

File: app/Http/Controllers/MappingController.php [hunk 1]
```diff
@@ -165,4 +170,56 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+
+    public function import(Webhook $webhook, MappingImportRequest $request)
+    {
+        DB::beginTransaction();
+        try {
+            $json = file_get_contents($request->file);
+            $getKeyValue = $this->mappingRepository->getKey($webhook);
+            $data = json_decode(file_get_contents($request->file), true);
+            $newData = [];
+            foreach($data as $key => $item) {
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:182]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.ControlStructures.ControlSignature.SpaceAfterKeyword</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 182</br>Expected 1 space(s) after FOREACH keyword; 0 found

File: app/Http/Controllers/MappingController.php [hunk 2]
```diff
@@ -165,4 +170,56 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+                if (!in_array($item['key'], $getKeyValue)) {
+                    array_push($newData, array_merge($item, ["webhook_id" => $webhook->id]));
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:184]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Strings.DoubleQuoteUsage.NotRequired</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 184</br>String "webhook_id" does not require double quotes; use single quotes instead

File: app/Http/Controllers/MappingController.php [hunk 3]
```diff
@@ -165,4 +170,56 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+                }
+            }
+
+            $this->mappingRepository->insert($newData);
+            DB::commit();
+            $request->session()->flash('messageSuccess', [
+                'status' => 'Upload file success',
+                'message' => 'This mapping successfully updaload file',
+            ]);
+
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:194]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.WhiteSpace.ControlStructureSpacing.SpacingBeforeClose</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 194</br>Blank line found at end of control structure

File: app/Http/Controllers/MappingController.php [hunk 4]
```diff
@@ -165,4 +170,56 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+        }
+    }
+
+    public function exportJson(Webhook $webhook, Request $request)
+    {
+        try {
+            $mapping = $this->mappingRepository->getKeyAndValue($webhook);
+            $fileName = time() . '_datafile.json';
+            $newJsonString = json_encode($mapping, JSON_PRETTY_PRINT);
+            File::put(public_path('/json/'.$fileName),$newJsonString);
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:210]: **Reporter**: CHECKSTYLE</br>**Rule**: Generic.Functions.FunctionCallArgumentSpacing.NoSpaceAfterComma</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 210</br>No space found after comma in argument list

File: app/Http/Controllers/MappingController.php [hunk 5]
```diff
@@ -165,4 +170,56 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+            $request->session()->flash('messageSuccess', [
+                'status' => 'export file success',
+                'message' => 'This mapping successfully export file',
+            ]);
+
+            return response()->download((public_path('/json/'. $fileName)));
+
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:217]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.WhiteSpace.ControlStructureSpacing.SpacingBeforeClose</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 217</br>Blank line found at end of control structure

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>failure</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

File: app/Http/Controllers/MappingController.php [hunk 6]
```diff
@@ -165,4 +170,55 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+    public function import(Webhook $webhook, MappingImportRequest $request)
+    {
+        DB::beginTransaction();
+        try {
+            $json = file_get_contents($request->file);
+            $getKeyValue = $this->mappingRepository->getKey($webhook);
+            $data = json_decode(file_get_contents($request->file), true);
+            $newData = [];
+
+            foreach($data as $key => $item) {
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:183]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.ControlStructures.ControlSignature.SpaceAfterKeyword</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 183</br>Expected 1 space(s) after FOREACH keyword; 0 found

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>failure</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

File: app/Http/Controllers/MappingController.php (hunk 6 above)
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:183]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.ControlStructures.ControlSignature.SpaceAfterKeyword</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 183</br>Expected 1 space(s) after FOREACH keyword; 0 found

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>failure</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

File: app/Http/Controllers/MappingController.php [hunk 7]
```diff
@@ -165,4 +170,54 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+
+    public function import(Webhook $webhook, MappingImportRequest $request)
+    {
+        DB::beginTransaction();
+        try {
+            $json = file_get_contents($request->file);
+            $getKeyValue = $this->mappingRepository->getKey($webhook);
+            $data = json_decode(file_get_contents($request->file), true);
+            $newData = [];
+            foreach($data as $key => $item) {
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:182]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.ControlStructures.ControlSignature.SpaceAfterKeyword</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 182</br>Expected 1 space(s) after FOREACH keyword; 0 found

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>failure</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

File: app/Http/Controllers/MappingController.php [hunk 8]
```diff
@@ -165,4 +169,54 @@ public function destroy(Webhook $webhook, Mapping $mapping)
             ]);
         }
     }
+
+    public function import(Webhook $webhook, MappingImportRequest $request)
+    {
+        DB::beginTransaction();
+        try {
+            $json = file_get_contents($request->file);
```
Comment (ghost) [inline app/Http/Controllers/MappingController.php]: cái $json này đang không dùng đến.

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

File: app/Http/Controllers/MappingController.php [hunk 9]
```diff
@@ -165,4 +169,54 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+            $getKeys = $this->mappingRepository->getKeys($webhook);
```
Comment (hatth-1632) [inline app/Http/Controllers/MappingController.php]:
```suggestion
            $keys = $this->mappingRepository->getKeys($webhook);
```

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

File: app/Http/Controllers/MappingController.php [hunk 10]
```diff
@@ -165,4 +169,54 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+                'message' => 'Import file failed. Something went wrong',
+            ]);
+        }
+    }
+
+    public function exportJson(Webhook $webhook, Request $request)
+    {
+        DB::beginTransaction();
+        try {
+            $getKeyAndValues = $this->mappingRepository->getKeyAndValues($webhook);
```
Comment (hatth-1632) [inline app/Http/Controllers/MappingController.php]:
```suggestion
            $keyValues = $this->mappingRepository->getKeyAndValues($webhook);
```

File: app/Http/Requests/MappingImportRequest.php [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'file' => 'required|file|mimetypes:application/json|max:1024',
```
Comment (hatth-1632) [inline app/Http/Requests/MappingImportRequest.php]: thiếu validate size

File: app/Repositories/Eloquents/MappingRepository.php [hunk 12]
```diff
@@ -11,4 +12,14 @@ public function getModel()
...
     }
+
+    public function getKeys($webhook)
+    {
+        return Mapping::where('webhook_id', $webhook->id)->pluck('key')->toArray();
+    }
+
+    public function getKeyAndValues($webhook)
+    {
+        return Mapping::where('webhook_id', $webhook->id)->select('key', 'value')->get();
```
Comment (hatth-1632) [inline app/Repositories/Eloquents/MappingRepository.php]:
```suggestion
        return Mapping::select('key', 'value')->where('webhook_id', $webhook->id)->get();
```

**Review by ghost — APPROVED** [review]

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>failure</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

File: app/Http/Controllers/MappingController.php [hunk 13]
```diff
@@ -165,4 +170,62 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
     }
+
+    public function import(Webhook $webhook, MappingImportRequest $request)
+    {
+        try {
+            $keys = $this->mappingRepository->getKeys($webhook);
+            $data = json_decode(file_get_contents($request->file), true);
+
+            Validator::make($data, [
+                "*.key" => "required|max:100",
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:181]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Strings.DoubleQuoteUsage.NotRequired</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 181</br>String "*.key" does not require double quotes; use single quotes instead

Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:181]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Strings.DoubleQuoteUsage.NotRequired</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 181</br>String "required|max:100" does not require double quotes; use single quotes instead

File: app/Http/Controllers/MappingController.php [hunk 14]
```diff
@@ -165,4 +170,62 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+                "*.value" => "required|max:100",
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:182]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Strings.DoubleQuoteUsage.NotRequired</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 182</br>String "*.value" does not require double quotes; use single quotes instead

Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:182]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Strings.DoubleQuoteUsage.NotRequired</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 182</br>String "required|max:100" does not require double quotes; use single quotes instead

File: app/Http/Controllers/MappingController.php [hunk 15]
```diff
@@ -165,4 +170,62 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+            ])->validate();
+
+            $newData = [];
+            foreach ($data as $item) {
+                if(in_array($item['key'], $keys)) {
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php:187]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.ControlStructures.ControlSignature.SpaceAfterKeyword</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 187</br>Expected 1 space(s) after IF keyword; 0 found

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

File: config/validation.php [hunk 16]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+return [
+    'validation' => [
+        'max_size_1m' => 1024, //MB
```
Comment (hatth-1632) [inline config/validation.php]:
```suggestion
        'max_size_1m' => 1024, //1MB
```

File: routes/web.php [hunk 17]
```diff
@@ -55,4 +55,7 @@
     Route::get('logs', '\Rap2hpoutre\LaravelLogViewer\LogViewerController@index')->middleware('admin');
     Route::put('templates/{template}/change_status', 'TemplateController@changeStatus');
     Route::resource('templates', 'TemplateController');
+    Route::post('webhooks/{webhook}/mappings/import', 'MappingController@import')->name('webhook.import.mappings');
+    Route::post('webhooks/{webhook}/mappings/exportJson', 'MappingController@exportJson')->name('webhook.export.mappings');
+    Route::get('webhooks/{webhook}/exports', 'MappingController@exportJson')->name('export-file');
```
Comment (hatth-1632) [inline routes/web.php:60]: 2 cái này cùng 1 hàm là sao em

File: app/Http/Controllers/MappingController.php [hunk 18]
```diff
@@ -165,4 +170,61 @@ public function destroy(Webhook $webhook, Mapping $mapping)
...
+                    return response()->json([
+                        'error' => true,
+                        'message' => "Key {$item['key']} exists.",
+                    ]);
+                }
+
+                array_push($newData, array_merge($item, ['webhook_id' => $webhook->id]));
+            }
+
+            $this->mappingRepository->insert($newData);
```
Comment (hatth-1632) [inline app/Http/Controllers/MappingController.php:198]: thêm DB transaction

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

**Pull request merged**